colossalai/utils/checkpoint/module_checkpoint.py

```diff
@@ -22,37 +22,52 @@ def save_checkpoint(dire: str,
         optimizer (ColossalaiOptimizer, optional): optimizers. Defaults to None.
         lr_scheduler (torch.optim.lr_scheduler._LRScheduler, optional): lr schedule. Defaults to None.
     """
-
-    mapping = dict()
-    new_dict = dict()
-
+    rank = dist.get_rank()
+    model_state = model.state_dict()
     # save the dist context about the tensors in a new dict, while still maintain the original dict.
-    for k, v in model.state_dict().items():
+    for k, v in model_state.items():
         if isinstance(v, ColoTensor):
-            mapping[k] = (v.dist_spec, v.compute_spec)
-            new_dict[k] = v.to_replicate().detach()
-        else:
-            new_dict[k] = v
-    if dist.get_rank() == 0:
-        for k, v in new_dict.items():
+            gather_tensor(v)  # gather shared tensors to rank0
+            # don't recover tensors in rank0, since the dict is only a copy of model
+
+    if rank == 0:
+        # sanity check
+        for k, v in model_state.items():
             if isinstance(v, ColoTensor):
+                assert v.save_ready
                 assert v.is_replicate()
+                delattr(v, 'save_ready')
+        # model saving
+        save_state = {'epoch': epoch, 'model': model_state}
```

Comment:
In the next PR. You can merge model and optim in a single file.
```
torch.save({
            'epoch': EPOCH,
            'model_state_dict': net.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'loss': LOSS,
            }, PATH)
```
like 
https://pytorch.org/tutorials/recipes/recipes/saving_and_loading_a_general_checkpoint.html